Implement UVarInt as a struct

freitagbr opened the pull request:
This implements `UVarInt` as a struct, which means that `UVarInt` instances are immutable. The `UVarInt` struct can be instantiated with several types and in two ways, now:
```crystal
# .new with any Int::Unsigned
a = UVarInt.new 10_u8
b = UVarInt.new 10_u32

#.new with any Enumerable(UInt8)
c = UVarInt.new [172_u8, 2_u8]
d = UVarInt.new 0x00_u8..0x09_u8
e = UVarInt.new "abcd"
# enumerable size must be less than 10
f = UVarInt.new(Array.new(20, 0_u8)) # raises ArgumentError

# [] macro, like Slice[]
g = UVarInt[172, 2] # g.uint == 300
```

Also implemented are `to_s`, `hexstring`, `to_i`, `to_u`, and basic arithmetic operations:
```
a = UVarInt[172, 2]
a.to_s 16 # => "12c"
a.hexstring # => "ac02"
b = UVarInt[172, 2]
c = a + b # c is a new UVarInt
c.to_i32 # 600_i32
c.to_u32 # 600_u32
```

freitagbr committed:
Implement UVarInt as a struct

Implement UVarInt as a struct, and provide several ways to instantiate:
from an Unsigned Integer, from Bytes, or from Enumerable(UInt8).
Provide accessors for the UVarInt both encoded and decoded.

freitagbr committed:
Refactor unit tests using new API

freitagbr committed:
Generalize encode to take Int::Unsigned

freitagbr committed:
Raise ArgumentError if more than 10 bytes passed

freitagbr committed:
Don't use globals, and declare values inline

freitagbr committed:
Add instantiation tests, reorganize

freitagbr committed:
Change Indexable initializer to Enumerable, add String initializer

freitagbr committed:
Add range and string test

freitagbr committed:
Move private methods up in scope

freitagbr committed:
Place fuzzing tests in describe

freitagbr committed:
Add macro for arithmetic operators and conversions

freitagbr committed:
Rename encoded -> bytes, decoded -> uint

freitagbr committed:
Add to_s and hexstring methods, [] macro

freitagbr committed:
Add tests for to_s, hexstring, and [] macro

freitagbr committed:
Make artihmetic operators return UVarInt

freitagbr commented:
Still need to add tests for arithmetic operators, and implement comparison operators.

nfischer commented on spec/uvarint_spec.cr:
```diff
@@ -1,77 +1,191 @@
+    it "raises an exception if passed more than 10 bytes" do
+      b = Array.new(20, 0_u8)
+      ex = expect_raises(ArgumentError) { UVarInt.new b }
+      ex.message.should eq("cannot initialize with more than 10 bytes")
```
Why can't we initialize with more than 10 bytes? Is this part of the standard, or is this an overflow issue? Shouldn't a varint instance ideally have no limit on size?

If this is an overflow issue when we try to access the integer value, then let's throw the exception upon decoding with something like `ArgumentError("This value overflows beyond __xx__ bytes")`.

freitagbr replied:
I guess it should be able to be instantiated with more than 10 bytes, but `instantiate` encodes the number immediately. Maybe we should store/return a `BigInt` if the size is larger than a `UInt64`.

nfischer replied:
Yeah, if BigInt is an option, let's use that. UVarint should be allowed to be arbitrarily large, otherwise the spec would just require Int64

nfischer commented on spec/uvarint_spec.cr:
```diff
@@ -1,77 +1,191 @@
+      n = Random.rand(0..0x7F).to_u8
+      b = Bytes[n]
+      v = UVarInt.new b
+      u = v.uint
```
`uint8` is the better method choice here, right?

freitagbr replied:
Yeah, `uint` and `bytes` are the accessors for the properties. Maybe it is better to just to `to_u8` and `to_bytes`.

nfischer replied:
Oh, sorry. I meant `to_u8` in my comment above. In that case, it's probably fine to leave as-is

nfischer commented on spec/uvarint_spec.cr:
```diff
@@ -1,77 +1,191 @@
+    it "returns self.bytes.hexstring" do
+      v = UVarInt.new 300_u64
+      h = v.hexstring
+      h.should eq("ac02")
```
Why does this differ from the case above? Shouldn't they both equal the same string?

freitagbr replied:
`hexstring` returns the hex representation of the bytes, whereas `to_s` returns the hex representation of the uint.

nfischer replied:
We should be careful here. It's very confusing if "Convert to a string, in base 16" and "Convert to a hexadecimal representation in string form" provide two different outcomes.

nfischer commented on spec/uvarint_spec.cr:
```diff
@@ -1,77 +1,191 @@
-      e = UVarint.encode n
-      d = UVarint.decode e
-      d.should eq(n)
+    it "encodes and decodes really big integers" do
```
Can you clarify what "really big" means here?

Also, can we add tests for numbers greater than 64 bit? This should work like:

 1. Instantiate a uvarint with a massive value (probably via a byte array or string)
 2. Serialize the uvarint to a string

freitagbr replied:
"Really big" just means a random `UInt64` between the max value of `UInt32` and `UInt64`. `Random.rand` only accepts up to an `Int32` for the range, so I had to generate the lower and upper 32 bits individually. I will clarify this in the description and perhaps with a comment.

If we add support for `BigInt`, we could add a test for that here too.

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+  end
+
+  def initialize(str : String)
+    raise ArgumentError.new "cannot initialize with more than 10 bytes" if str.size > 10
```
I don't think we should raise this ArgumentError here, but we should instead raise an ArgumentError if this string contains non-hex characters.

freitagbr replied:
I don't the the string is treated as a hex reperesentation, rather the numeric ASCII value of each character is used. Maybe we should instead treat the string as a hex?

nfischer replied:
I was under the impression that `UVarint("30")` was equivalent to `UVarint(48)`, because `0x30` in hex is 48 in decimal. Am I misunderstanding? What is the equivalent of `UVarint(48)` and why?

freitagbr replied:
The ascii code points are used, so `"3"` is `51` in dec of `0x33` in hex, `"0"` is `48` in dec or `0x30` in hex. So, `"30"` is `0x3330` in hex, which is `13104`. In other words, `UVarInt("30") == UVarInt(13104)`. I think treating the string as a hex representation makes more sense though.

nfischer replied:
If we have a `.to_s` method, I only see problems. `UVarInt(34).to_s` should produce `"34"`, but this isn't the result we want for serializing. On the other hand, if someone is trying to serialize a VarInt, they'll likely try `.to_s` first and be bitten by the wrong result. Best thing is probably to leave off `.to_s` entirely and instead go with `.encode()`. It's probably ok to leave `.to_i`.

If someone really wants `"34"` from a UVarInt, we could recommend:

```crystal
u = UVarInt(34)
...
u.to_i64.to_s # invokes Int64's .to_s method instead
```

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+end
+
+struct UVarInt
+  @bytes : Bytes
```
Why are we storing both the `@bytes` and `@uint`? The `UInt64` is a poor choice because it has an inherent limit, which is what uvarint is supposed to avoid. Also, if both these members are supposed to represent the same information, then keeping both may likely lead to inconsistent state.

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+  end
+
+  def to_s(base : Int32)
+    @uint.to_s base
```
Ideally, `.to_s` should not be limited by `UInt64`'s `.to_s` implementation. This isn't critical as long as `hexstring` works on arbitrary sizes, though, as long as this limitation is clearly documented.

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+  end
+
+  macro [](*args)
+    {% if args.size > 10 %}
```
Also, if we really do need to keep `10` hardcoded for some reason, we should move this into a constant. But hopefully we don't need it at all.

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+  end
+
+  {% begin %}
+    {% for opt in %w(% * ** + - / << <=> === >>) %}
```
Could you add some comments explaining this code? Also, could you add test cases?

Do we want `+=` et al.? If so, then this should be a class, not a struct. We can consider adding these later as well.

Also, what about `&&` and `||`? And `!`?

nfischer replied:
Also, in general, please add tests for a couple operators.

freitagbr replied:
Yeah, this is some code that should be documented. I will add tests for every operator as well, and add comparison operators.

As for the reassignment operators, I'll have to think about it.

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+    {% end %}
+
+    {% for name, type in {
+        to_i: Int32, to_u: UInt32, to_f: Float64,
```
nit: Could we improve readability by wrapping the pairs in parentheses or maybe aligning columns?

freitagbr replied:
I don't think parens will work here, but yes, I will align the columns.

nfischer commented on src/uvarint.cr:
```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
+      %bytes = Bytes[{{*args}}]
+      UVarInt.new %bytes
+    {% end %}
+  end
```
Does this `end` go here? I thought it went at the end of all the macros. Am I misunderstanding something?

freitagbr replied:
For macros defined with `macro`, `end` does indeed go there. Check out the source for [`Slice`](https://github.com/crystal-lang/crystal/blob/7f82f79bd1e3a7a1a73607e35c0662906736f1f8/src/slice.cr#L28) for example.

nfischer replied:
What I mean is, how can you use macro syntax outside of the `macro ... end` block, for the macros below?

freitagbr committed:
Columnize names and types

freitagbr committed:
Reimplement with BigInt, remove limit on size

freitagbr committed:
Add BigInt initializer and remove bounds checks

freitagbr committed:
Add BigInt test and remove overflow tests

freitagbr committed:
Remove 10 argument limit from [] macro

freitagbr committed:
Change @uint -> @bigint and add macro comments

freitagbr committed:
Use #to_big_i instead of @uint accessor in tests

freitagbr committed:
Add logic operators

freitagbr committed:
Treat strings during construct as hex

freitagbr commented:
Ok, I addressed most of your comments:

* implemented all of the math and logic operators, and added tests for them
* changed the internal represenation of the unsigned integer to a `BigInt` and removed all length limits
* changed the `String` constructor to treat the string as a hexadecimal string

There remains the difference between `#to_s` and `#hexstring`, where the former returns a string representation of the stored number, and `#hexstring` returns a hexadecimal string representation of the stored bytes. The `Bytes` has a `#hexstring` method (that's where `UVarInt#hexstring` comes from), so maybe this could be removed from `UVarInt`, and let it be accessed through `UvarInt.bytes#hexstring`.

freitagbr commented:
@nfischer Just pinging you about this PR.

nfischer commented on src/uvarint.cr:
```diff
@@ -1,44 +1,131 @@
+
+  def initialize(int : BigInt)
+    @bigint = int.to_big_i
+    @bytes = encode int
```
It's a bad idea to store both the bigint *and* the bytes. Pick one, otherwise we're doing twice the work and introducing the possibility for bugs (inconsistent state).

freitagbr replied:
Structs are immutable though, right? Once they're set, they cannot be changed. I prefer it this way, because we don't have to recalculate one of them every time we want the number or we want the bytes.

nfischer commented on src/uvarint.cr:
```diff
@@ -1,44 +1,131 @@
+  end
+
+  def to_s(base : Int32)
+    @bigint.to_s base
```
If `to_s` doesn't return the serialized value, we should cut it out. Keeping it around is just confusing for users.

I think we should add a `.serialize()` method which returns the serialized value. Much clearer about what it does.

freitagbr replied:
Well, a `UVarInt` just represents a number. All of the number types in crystal have a `to_s` method that returns the string representation of the number, so I figured the `to_s` method here should do the same.

For the string representation of the bytes, there is the `hexstring` method, which is inherited from the `Slice` type.

nfischer commented on src/uvarint.cr:
```diff
@@ -1,44 +1,131 @@
+  arr = [] of UInt8
+  while n >= MSB
+    arr << (n.to_u8 | MSB)
+    n = n >> 7
```
Prefer named constants over magic numbers like 7

nfischer commented on src/uvarint.cr:
```diff
@@ -1,44 +1,131 @@
-    end
-    return 0_u64
+    x = x | ((b & REST).to_big_i << s)
+    s += 7
```
Magic numbers

freitagbr replied:
I agree, they're magic, and I do not fully understand them. I used named constants where I could (`REST` for example). I left a comment at the top of the file with a link to the golang implementation, so if there are questions about the implement, a user could visit that link.

nfischer replied:
@freitagbr the magic isn't so obscure. I'm pretty sure we use 7 because there's 8 bits in a byte. If you shiftright by 7, you get MSB. If you shift left by 7 again, you get MSB followed by all zeros. I haven't reviewed the algorithm in a bit, but I imagine that's the use case.

nfischer commented on src/uvarint.cr:
```diff
@@ -1,44 +1,131 @@
-      n = n >> 7
+# Decodes enumerable bytes into a BigInt.
+# If the enumeration is longer thar 10 bytes,
+# then an overflow exception is raised.
```
Why are we having an overflow exception? Is this the limit of the BigInt type?

freitagbr replied:
This is no longer the case, I forgot to remove this part of the comment.

nfischer commented on src/uvarint.cr:
```diff
@@ -1,44 +1,131 @@
+# then an overflow exception is raised.
+private def decode(bytes : Bytes) : BigInt
+  x = ZERO
+  s = 0
```
This part looks kind of unusual. We use `ZERO` on one line, and 0 on the other. If this is for type reasons, it might be worth adding a comment.

freitagbr replied:
I agree, that is confusing. For reference, `ZERO == BigInt(0)`. What would be a better name here? `x` is the `BigInt` value that is decoded into.

nfischer commented:
@freitagbr we've been stretching this PR out for quite a while. I'll just mark these changes as approved now, feel free to address comments down the road.

freitagbr committed:
Fix comments

freitagbr committed:
Remove BigInt casting from BigInt constructor
